File: sdk-py/src/proto402/signing.py
"""ed25519 signing helpers for 402proto."""
from __future__ import annotations

import based58
from nacl.signing import SigningKey, VerifyKey


def sign(canonical_body: bytes, secret_key_bytes: bytes) -> str:
    """sign canonical body. returns base58-encoded signature."""
    signer = SigningKey(secret_key_bytes[:32])
    sig = signer.sign(canonical_body).signature
    return based58.b58encode(sig).decode("ascii")


def verify(canonical_body: bytes, signature_b58: str, pubkey_b58: str) -> bool:
    """verify a 402proto signature. returns True on success, False otherwise."""
    from nacl.exceptions import BadSignatureError

    try:
        sig = based58.b58decode(signature_b58.encode("ascii"))
        pk = based58.b58decode(pubkey_b58.encode("ascii"))
        verifier = VerifyKey(pk)
        verifier.verify(canonical_body, sig)
        return True
    except (BadSignatureError, ValueError):
        return False


def derive_pubkey(secret_key_bytes: bytes) -> str:
    """derive base58 public key from ed25519 secret."""
    signer = SigningKey(secret_key_bytes[:32])
    return based58.b58encode(bytes(signer.verify_key)).decode("ascii")

File: sdk-ts/src/signing.ts
/**
 * ed25519 signing helpers (noble curves).
 */
import { ed25519 } from "@noble/curves/ed25519";
import bs58 from "bs58";

export function sign(canonicalBody: Uint8Array, secretKey: Uint8Array): string {
  // noble takes raw 32-byte seed
  const seed = secretKey.length >= 32 ? secretKey.slice(0, 32) : secretKey;
  const sig = ed25519.sign(canonicalBody, seed);
  return bs58.encode(sig);
}

export function verify(canonicalBody: Uint8Array, signatureB58: string, pubkeyB58: string): boolean {
  try {
    const sig = bs58.decode(signatureB58);
    const pk = bs58.decode(pubkeyB58);
    return ed25519.verify(sig, canonicalBody, pk);
  } catch {
    return false;
  }
}

export function derivePubkey(secretKey: Uint8Array): string {
  const seed = secretKey.length >= 32 ? secretKey.slice(0, 32) : secretKey;
  const pub = ed25519.getPublicKey(seed);
  return bs58.encode(pub);
}
